Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -939,26 +977,398 @@ Query Serializer::DecodeQueryTarget(
+  FilterList result;
+
+  switch (proto.which_filter_type) {
+    case google_firestore_v1_StructuredQuery_Filter_composite_filter_tag:
+      return DecodeCompositeFilter(reader, proto.composite_filter);
+
+    case google_firestore_v1_StructuredQuery_Filter_unary_filter_tag:
+      return result.push_back(DecodeUnaryFilter(reader, proto.unary_filter));
+
+    case google_firestore_v1_StructuredQuery_Filter_field_filter_tag:
+      return result.push_back(DecodeFieldFilter(reader, proto.field_filter));
+
+    default:
+      reader->Fail(StringFormat("Unrecognized Filter.which_filter_type %s",
+                                proto.which_filter_type));
+      return result;
+  }
+}
+
+google_firestore_v1_StructuredQuery_Filter Serializer::EncodeNonCompositeFilter(
+    const FieldFilter& filter) const {
+  google_firestore_v1_StructuredQuery_Filter result{};
+
+  if (filter.op() == Filter::Operator::Equal) {
+    if (filter.value().is_null() || filter.value().is_nan()) {
+      result.which_filter_type =
+          google_firestore_v1_StructuredQuery_Filter_unary_filter_tag;
+      result.unary_filter.which_operand_type =
+          google_firestore_v1_StructuredQuery_UnaryFilter_field_tag;
+      result.unary_filter.field.field_path = EncodeFieldPath(filter.field());
+
+      auto op =
+          filter.value().is_null()
+              ? google_firestore_v1_StructuredQuery_UnaryFilter_Operator_IS_NULL
+              : google_firestore_v1_StructuredQuery_UnaryFilter_Operator_IS_NAN;
+      result.unary_filter.op = op;
+
+      return result;
+    }
   }
 
-  // TODO(rsgowman): Decode the filters.
-  // TODO(rsgowman): Decode the orders.
-  // TODO(rsgowman): Decode the limit.
-  // TODO(rsgowman): Decode the startat.
-  // TODO(rsgowman): Decode the endat.
+  result.which_filter_type =
+      google_firestore_v1_StructuredQuery_Filter_field_filter_tag;
+
+  result.field_filter.field.field_path = EncodeFieldPath(filter.field());
+  result.field_filter.op = EncodeFieldFilterOperator(filter.op());
+  result.field_filter.value = EncodeFieldValue(filter.value());
 
-  return Query(path);
+  return result;
+}
+
+Filter Serializer::DecodeFieldFilter(
+    nanopb::Reader* reader,
+    const google_firestore_v1_StructuredQuery_FieldFilter& field_filter) const {
+  FieldPath field_path =
+      FieldPath::FromServerFormat(DecodeString(field_filter.field.field_path));
+  Filter::Operator op = DecodeFieldFilterOperator(reader, field_filter.op);
+  FieldValue value = DecodeFieldValue(reader, field_filter.value);
+
+  return FieldFilter::Create(std::move(field_path), op, std::move(value));
+}
+
+Filter Serializer::DecodeUnaryFilter(
+    nanopb::Reader* reader,
+    const google_firestore_v1_StructuredQuery_UnaryFilter& unary) const {
+  if (unary.which_operand_type !=
+      google_firestore_v1_StructuredQuery_UnaryFilter_field_tag) {
+    reader->Fail(StringFormat("Unexpected UnaryFilter.which_operand_type: %s",
```

Comment:
This seems like it should be an assertion since this is a violation of the precondition of this call.

Contrast with the failure in `DecodeCompositeFilter`, which I think is appropriate since nothing else is checking that the operator is AND.

Reply:
Done, thanks.